tests/e2e/websocket/websocket_ssl.cpp

```diff
@@ -0,0 +1,255 @@
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+#include <gflags/gflags.h>
+#include <algorithm>
+#include <chrono>
+#include <cstdint>
+#include <iostream>
+#include <memory>
+#include <string>
+#include <string_view>
+#include <thread>
+#include <utility>
+#include <vector>
+
+#include <fmt/core.h>
+#include <spdlog/spdlog.h>
+#include <unistd.h>
+
+#include "common.hpp"
+#include "utils/logging.hpp"
+
+DEFINE_uint64(bolt_port, 7687, "Bolt port");
+
+class Session : public std::enable_shared_from_this<Session> {
+ public:
+  explicit Session(net::io_context &ioc, ssl::context &ctx, std::vector<std::string> &expected_messages)
+      : resolver_(net::make_strand(ioc)), ws_(net::make_strand(ioc), ctx), received_messages_{expected_messages} {}
+
+  explicit Session(net::io_context &ioc, ssl::context &ctx, std::vector<std::string> &expected_messages,
+                   Credentials creds)
+      : resolver_(net::make_strand(ioc)),
+        ws_(net::make_strand(ioc), ctx),
+        received_messages_{expected_messages},
+        creds_{creds} {}
+
+  void Run(std::string host, std::string port) {
+    host_ = host;
+    resolver_.async_resolve(host, port, beast::bind_front_handler(&Session::OnResolve, shared_from_this()));
+  }
+
+  void OnResolve(beast::error_code ec, tcp::resolver::results_type results) {
+    if (ec) {
+      return Fail(ec, "resolve");
+    }
+
+    beast::get_lowest_layer(ws_).expires_after(std::chrono::seconds(30));
+
+    beast::get_lowest_layer(ws_).async_connect(results,
+                                               beast::bind_front_handler(&Session::OnConnect, shared_from_this()));
+  }
+
+  void OnConnect(beast::error_code ec, tcp::resolver::results_type::endpoint_type ep) {
+    if (ec) {
+      return Fail(ec, "connect");
+    }
+
+    host_ = fmt::format("{}:{}", host_, ep.port());
+
+    beast::get_lowest_layer(ws_).expires_after(std::chrono::seconds(30));
+
+    if (!SSL_set_tlsext_host_name(ws_.next_layer().native_handle(), host_.c_str())) {
+      ec = beast::error_code(static_cast<int>(::ERR_get_error()), net::error::get_ssl_category());
+      return Fail(ec, "connect");
+    }
+
+    ws_.next_layer().async_handshake(ssl::stream_base::client,
+                                     beast::bind_front_handler(&Session::OnSSLHandshake, shared_from_this()));
+  }
+
+  void OnSSLHandshake(beast::error_code ec) {
+    if (ec) {
+      return Fail(ec, "ssl_handshake");
+    }
+
+    beast::get_lowest_layer(ws_).expires_never();
+    ws_.set_option(websocket::stream_base::timeout::suggested(beast::role_type::client));
+    ws_.set_option(websocket::stream_base::decorator([](websocket::request_type &req) {
+      req.set(http::field::user_agent, std::string(BOOST_BEAST_VERSION_STRING) + " websocket-client-async-ssl");
+    }));
+
+    ws_.async_handshake(host_, "/", beast::bind_front_handler(&Session::OnHandshake, shared_from_this()));
+  }
```

Comment:
Maybe consider executing this conditionally and in this way the two clients might be merged into one.